hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/replication/VerifyReplicationRecompareRunnable.java

```diff
@@ -0,0 +1,156 @@
+    Scan tableScan, Table sourceTable, Table replicatedTable, int reCompareTries,
+    int sleepMsBeforeReCompare, int reCompareBackoffExponent, boolean verbose) {
+    this.context = context;
+    this.sourceResult = sourceResult;
+    this.replicatedResult = replicatedResult;
+    this.originalCounter = originalCounter;
+    this.delimiter = delimiter;
+    this.tableScan = tableScan;
+    this.sourceTable = sourceTable;
+    this.replicatedTable = replicatedTable;
+    this.reCompareTries = reCompareTries;
+    this.sleepMsBeforeReCompare = sleepMsBeforeReCompare;
+    this.reCompareBackoffExponent = reCompareBackoffExponent;
+    this.verbose = verbose;
+    this.row = VerifyReplication.getRow(sourceResult, replicatedResult);
+  }
+
+  @Override
+  public void run() {
+    Get get = new Get(row);
+    get.setCacheBlocks(tableScan.getCacheBlocks());
+    get.setFilter(tableScan.getFilter());
+
+    int sleepMs = sleepMsBeforeReCompare;
+    int tries = 0;
+
+    while (++tries <= reCompareTries) {
+      context.getCounter(VerifyReplication.Verifier.Counters.RE_COMPARES).increment(1);
+
+      try {
+        Thread.sleep(sleepMs);
+      } catch (InterruptedException e) {
+        LOG.warn("Sleeping interrupted, incrementing bad rows and aborting");
+        incrementOriginalAndBadCounter();
+        Thread.currentThread().interrupt();
+      }
+
+      try {
+        if (fetchLatestRows(get)) {
+          if (matches(sourceResult, replicatedResult, null)) {
+            if (verbose) {
+              LOG.info("Good row key (with recompare): {}{}{}", delimiter,
+                Bytes.toStringBinary(row), delimiter);
+            }
+            context.getCounter(VerifyReplication.Verifier.Counters.GOODROWS).increment(1);
+            return;
+          }
+        }
+      } catch (IOException e) {
+        context.getCounter(VerifyReplication.Verifier.Counters.FAILED_RECOMPARE).increment(1);
+        if (verbose) {
+          LOG.info("Got an exception during recompare for rowkey={}", Bytes.toStringBinary(row), e);
+        }
+      }
+
+      sleepMs = sleepMs * (2 ^ reCompareBackoffExponent);
+    }
+
+    LOG.error("{}, rowkey={}{}{}", originalCounter, delimiter, Bytes.toStringBinary(row),
+      delimiter);
+    incrementOriginalAndBadCounter();
+  }
+
+  private boolean fetchLatestRows(Get get) throws IOException {
+    Result sourceResult = sourceTable.get(get);
+    Result replicatedResult = replicatedTable.get(get);
+
+    boolean sourceMatches = matches(sourceResult, this.sourceResult,
+      VerifyReplication.Verifier.Counters.SOURCE_ROW_CHANGED);
+    boolean replicatedMatches = matches(replicatedResult, this.replicatedResult,
+      VerifyReplication.Verifier.Counters.PEER_ROW_CHANGED);
+
+    if (sourceMatches && replicatedMatches) {
+      return true;
+    }
+
+    this.sourceResult = sourceResult;
+    this.replicatedResult = replicatedResult;
+    return false;
```

Comment:
not a huge deal, but could this be simplifed to:

```java
this.sourceResult = sourceResult;
this.replicatedResult = replicatedResult;
return sourceMatches && replicatedMatches;
```

not sure there's a huge reason to only update the `this.` values if no match? correct me if wrong

Reply:
yeah you're correct, if the values match then `this.sourceResult = sourceResult` and `this.replicatedResult = replicatedResult` are effectively no-ops, so I can definitely simplify here